src/vcon/MarketGovernance.sol

```diff
@@ -0,0 +1,653 @@
+    function _movePCVWithChecks(
+        address source,
+        address destination,
+        address swapper,
+        uint256 amountPcv,
+        uint256 totalPcv,
+        int256 sourceVenueBalance,
+        int256 destinationVenueBalance,
+        uint256 totalVconStaked
+    ) private {
+        address sourceAsset = IPCVDepositV2(source).balanceReportedIn();
+        address destinationAsset = IPCVDepositV2(destination)
+            .balanceReportedIn();
+
+        /// validate pcv movement
+        /// check underlying assets match up and if not that swapper is provided and valid
+        PCVRouter(pcvRouter).movePCV(
+            source,
+            destination,
+            swapper,
+            amountPcv,
+            sourceAsset,
+            destinationAsset
+        );
+
+        /// if nothing is staked on source, ignore balance check
+        if (sharesToVcon(source, venueTotalShares[source]) > 0) {
+            int256 sourceVenueBalanceAfter = getVenueDeviation(
+                source,
+                totalPcv,
+                totalVconStaked
+            );
+
+            /// source and dest venue balance measures the distance from being perfectly balanced
+
+            /// validate source venue balance became more balanced
+            _checkBalance(
+                sourceVenueBalance,
+                sourceVenueBalanceAfter,
+                "MarketGovernance: src more imbalanced"
+            );
+        }
+
+        int256 destinationVenueBalanceAfter = getVenueDeviation(
+            destination,
+            totalPcv,
+            totalVconStaked
+        );
+
+        /// validate destination venue balance became more balanced
+        _checkBalance(
+            destinationVenueBalance,
+            destinationVenueBalanceAfter,
+            "MarketGovernance: dest more imbalanced"
+        );
+    }
+
+    /// @notice helper function to validate balance moved in the right direction after a pcv movement
+    /// if balance before and after are the same, return true
+    function _checkBalance(
+        int256 balanceBefore,
+        int256 balanceAfter,
+        string memory reason
+    ) private pure {
+        require(
+            balanceBefore < 0 /// if balance is under weight relative to vcon staked, ensure it doesn't go over balance
+                ? balanceAfter >= balanceBefore && balanceAfter <= 0 /// if balance is over weight relative to vcon staked, ensure it doesn't go under balance
+                : balanceAfter <= balanceBefore && balanceAfter >= 0,
+            reason
+        );
+    }
+
+    /// @notice update the venue last recorded profit
+    /// and the venue last recorded vcon share price
+    /// system must be at lock level 1 to call this function,
+    /// otherwise call to `accrue()` on the PCV Deposit will fail
+    /// @dev this function assumes the venue is in the PCV Oracle as the
+    /// calling function should either check, or be callable only by
+    /// governance
```

Comment:
the reason I don't want to do this is it would require a reference to another smart contract, and we don't want to allow reentrancy in this contract even though we follow CEI.

Reply:
we also call withdraw on PCV Deposits via the PCV Router, so we need to get into level 1 locked state before calling that function